yt/yt/server/node/exec_node/job_proxy_log_manager.cpp

```diff
@@ -0,0 +1,233 @@
+
+static const auto& Logger = ExecNodeLogger;
+
+////////////////////////////////////////////////////////////////////////////////
+
+class TJobProxyLogManager
+    : public IJobProxyLogManager
+{
+public:
+    TJobProxyLogManager(IBootstrap* bootstrap)
+        : Bootstrap_(bootstrap)
+        , Config_(Bootstrap_->GetConfig()->ExecNode->JobProxyLogManager)
+        , Directory_(Config_->Directory)
+        , ShardingKeyLength_(Config_->ShardingKeyLength)
+        , LogsStoragePeriod_(Config_->LogsStoragePeriod)
+        , DirectoryTraversalConcurrency_(Config_->DirectoryTraversalConcurrency)
+        , AsyncSemaphore_(New<NConcurrency::TAsyncSemaphore>(Config_->DirectoryTraversalConcurrency.value_or(0)))
+    {}
+
+    void Start() override
+    {
+        Bootstrap_->GetStorageHeavyInvoker()->Invoke(BIND(
+            [this, this_ = MakeStrong(this)] {
+                CreateShardingDirectories();
+                TraverseJobDirectoriesAndScheduleRemovals();
+            }));
+    }
+
+    void OnJobUnregistered(TJobId jobId) override
+    {
+        NConcurrency::TDelayedExecutor::Submit(
+            BIND(&TJobProxyLogManager::RemoveJobLog, MakeStrong(this), jobId),
+            LogsStoragePeriod_,
+            Bootstrap_->GetStorageHeavyInvoker()
+        );
+    }
+
+    TString GetShardingKey(TJobId jobId) override
+    {
+        auto entropy = NScheduler::EntropyFromAllocationId(
+            NScheduler::AllocationIdFromJobId(jobId));
+        auto entropyHex = Format("%016lx", entropy);
+        return entropyHex.substr(0, ShardingKeyLength_);
+    }
+
+    void OnDynamicConfigChanged(
+        TJobProxyLogManagerDynamicConfigPtr /*oldConfig*/,
+        TJobProxyLogManagerDynamicConfigPtr newConfig) override
+    {
+        LogsStoragePeriod_ = newConfig->LogsStoragePeriod;
+        DirectoryTraversalConcurrency_ = newConfig->DirectoryTraversalConcurrency;
+        AsyncSemaphore_->SetTotal(DirectoryTraversalConcurrency_.value_or(0));
+    }
+
+    void SaveJobProxyLog(
+        TJobId jobId,
+        const NYPath::TYPath& outputPath,
+        NObjectClient::TTransactionId transactionId) override
+    {
+        auto logsPath = JobIdToLogsPath(jobId);
+
+        YT_LOG_INFO("Requested saving logs for path %v to ypath %v, transactionId %v", logsPath, outputPath, transactionId);
+
+        auto logFile = TFile(NFS::CombinePaths(logsPath, "job_proxy.log"), OpenExisting | RdOnly);
+        auto logFileLength = logFile.GetLength();
+
+        auto buffer = TSharedMutableRef::Allocate(logFile.GetLength());
+        logFile.Read((void*)buffer.Data(), logFileLength);
+
+        NApi::TFileWriterOptions options;
+        options.TransactionId = transactionId;
+        auto writer = Bootstrap_->GetClient()->CreateFileWriter(outputPath, options);
+
+        NConcurrency::WaitFor(writer->Open())
+            .ThrowOnError();
+
+        NConcurrency::WaitFor(writer->Write(buffer))
```

Comment:
It is better to read file gradually withing some small buffer (1_MB for example) and push this data to writerr

Reply:
Added a buffer with configurable size.